pdns/iputils.hh

```diff
@@ -369,6 +370,307 @@ private:
+        delete node->node4;
+        node->node4 = NULL;
+      } else {
+        delete node->node6;
+        node->node6 = NULL;
+      }
+    }
+  }
+
+  void erase(const string &mask) {
+    erase(Netmask(mask));
+  }
+
+  node_type* insert(const string &mask) {
+    return insert(Netmask(mask));
+  }
+
+  //<! Creates new value-pair in tree and returns it.
+  node_type* insert(const Netmask &mask) {
+    // lazily initialize tree on first insert.
+    if (root == NULL) root = new TreeNode(0);
+    TreeNode* node = root;
+    node_type* value = NULL;
+
+    if (mask.getNetwork().sin4.sin_family == AF_INET) {
+      std::bitset<32> addr(ntohl(mask.getNetwork().sin4.sin_addr.s_addr));
+      int bits = 0;
+      // we turn left on 0 and right on 1
+      while(bits < mask.getBits()) {
+        uint8_t val = addr[31-bits];
+        if (val)
+          node = node->make_right();
+        else
+          node = node->make_left();
+        bits++;
+      }
+      // only create node if not yet assigned
+      if (node->node4 == NULL) node->node4 = new node_type();
+      value = node->node4;
+    } else {
+      uint64_t* addr = (uint64_t*)mask.getNetwork().sin6.sin6_addr.s6_addr;
+      std::bitset<64> addr_low(be64toh(addr[1]));
+      std::bitset<64> addr_high(be64toh(addr[0]));
+      int bits = 0;
+      while(bits < mask.getBits()) {
+        uint8_t val;
+        // we use high address until we are
+        if (bits < 64) val = addr_high[63-bits];
+        // past 64 bits, and start using low address
+        else val = addr_low[127-bits];
+
+        // we turn left on 0 and right on 1
+        if (val)
+          node = node->make_right();
+        else
+          node = node->make_left();
+        bits++;
+      }
+      // only create node if not yet assigned
+      if (node->node6 == NULL) node->node6 = new node_type();
+      value = node->node6;
+    }
+    // assign mask
+    value->first = mask;
+    // keep a copy of all nodes, handy later.
+    _nodes.push_back(value);
+
+    return value;
+  }
+
+  const node_type* lookup(const ComboAddress& value) const {
+    TreeNode *node = root;
+    TreeNode *last = NULL;
+    node_type *ret = NULL;
+
+    // no tree, no value
+    if ( node == NULL ) return NULL;
+
+    // exact same thing as above, except
+    if (value.sin4.sin_family == AF_INET) {
+      std::bitset<32> addr(ntohl(value.sin4.sin_addr.s_addr));
+      int bits = 0;
+      while(bits < 32) {
+        // ...we keep track of last non-empty node
+        if (node->node4) last = node;
+        uint8_t val = addr[31-bits];
+        // ...and we don't create left/right hand
+        if (val) {
+          if (node->right) node = node->right;
+          // ..and we break when road ends
+          else break;
+        } else {
+          if (node->left) node = node->left;
+          else break;
+        }
+        bits++;
+      }
+      // use last found as node if necessary
```

Comment:
sorry; just the two following lines trying to find `res` which you already found above.